### ISSUE TITLE
Performance issue generating Squel queries

### ISSUE THREAD

**Issue by connor4312** [hiddentao/squel#257]
We were doing some load testing on our servers and determined that building queries with Squel was causing performance issues. Particularly, one of the more complex queries was taking 4ms to build from Squel -- that's 4ms on which the event loop is blocked and nothing else can occur. For an API where we like to have <8ms Node-time per request, that is a huge chunk of time.

Upon further investigation this appears to be a structural issue with Squel, particularly its use of ES6 classes and `super()`. When transpiled to ES5 these are quite slow, and still rather slow (only 3-4x faster) natively in Node. These calls make even the simplest of queries take a notable amount of time to build. On Node 6, for example, running against the still somewhat-slow knex:

```js
const squel = require('./');
const knex = require('knex')({});

bench('squel', () => {
    squel.select().from('students').toString();
});
bench('knex', () => {
    knex('students').select().toString();
});

/*
➜  squel git:(master) ✗ matcha benchmark.js

           1,050 op/s » squel
          22,483 op/s » knex

  Suites:  0
  Benches: 2
  Elapsed: 2,215.62 ms
*/
```

Fixing this issue is probably outside the scope of some random PR, but we wanted to make you guys aware of the issue going forward. As implementations of ES6 mature Squel will likely become faster, but this is likely to be an ongoing concern for high-throughput applications.

**Comment by hiddentao** [comment]
Thanks for bringing this up. My long-term stance on  performance has been that other parts of people apps would most likely have a larger impact on performance than the query building parts, and hence it wasn't something I needed to particularly focus on. 

But those numbers are quite shocking! I'm going to add this to my todo list for my next Squel sprint.

**Comment by connor4312** [comment]
Great, let me know if there's anything you want help/PRs for 😄

**Comment by wirehead** [comment]
Yeah, I love how squel works.. but I've been noticing that it's eating an uncomfortably large amount of CPU as well.

One suggestion is making it easier to curry the query.  

In the codebase I'm using squel in, I've got some cases where I'm using squel to generate arbitrary queries with a bunch of different parameters.  Those are a bit harder to tune.  I'll probably just wrap them in a cache.

But I've also got some cases where I'm just using squel to generate the exact same query with different parameters, where the only thing that changes is the database entity I'm trying to look up.  An example PR where I hack around this is rm3web/rm3#329 -- I cache the generated query SQL string and then substitute different parameters.  It's kinda ugly.  It would be cool to have a slightly more friendly way inside of squel to specify placeholder parameters and get a re-usable function.

**Comment by schmod** [comment]
If this is still an issue, give 3.9.1 a try.  #309 and #310 introduced a few optimizations that should benefit large queries.

**Comment by connor4312** [comment]
Both squel and knex seem to have gotten a little faster, but squel still trails by a fair margin.

```
➜  squel-test node --version
v6.4.0
➜  squel-test matcha bench.js
(node:44648) DeprecationWarning: child_process: options.customFds option is deprecated. Use options.stdio instead.

           2,259 op/s » squel
          55,135 op/s » knex

  Suites:  0
  Benches: 2
  Elapsed: 1,796.55 ms
```

For our application we made the hop over to knex internally, so this is no longer an ongoing issue for us. (Though I personally. would still be happy to lend a hand for specific performance if desired!)

**Comment by schmod** [comment]
Just curious: What's the script you're using to test?

**Comment by connor4312** [comment]
The script I'm using is up at the top of the issue. I'm running it with [matcha](https://github.com/logicalparadox/matcha)

**Comment by brandonros** [comment]
Is there anything out there faster than knex?

**Comment by brandonros** [comment]
What would it take to get this library to not totally suck performance wise?

**Comment by connor4312** [comment]
If you're interested in introducing some PRs to make it "not totally suck", I'm sure the maintainers would welcome them, otherwise you are free to conduct discourse in a more positive and friendly way 🙂 

Since I'm here, another run with the same code:

```
➜  foo node --version
v8.9.0
➜  foo matcha test.js
(node:34410) [DEP0006] DeprecationWarning: child_process: options.customFds option is deprecated. Use options.stdio instead.

           3,993 op/s » squel
          84,338 op/s » knex

  Suites:  0
  Benches: 2
  Elapsed: 1,665.12 ms
```

**Comment by hiddentao** [comment]
@brandonros @connor4312 Yep, performance isn't as good as knex, mainly I think because of the inefficiencies of the Babel-translated ES6 code. Truth is I don't get a lot of time to work on this library nowadays since I first of all don't even use SQL dbs at the moment. I still do fix bugs and make improvements every now and then, and of course am happy to accept PRs for the same.

**Comment by brandonros** [comment]
Could you point me in the right direction of where to start to fix it?

Regards,

Brandon
________________________________

**Comment by brandonros** [comment]
I took it into the liberty of my own hands to eject my *production* codebase that makes hundreds of thousands of queries a night, all built by this squel package, to move to a package with 20x better performance: knex.

I did not want to rewrite all of my tests + code for the mildly different syntax difference, so I made this:

```
var knex = require('knex')({
  client: 'mysql',
  wrapIdentifier: function(value) {
    return value;
  }
});

var QueryBuilder = function() {
  this.operation = null;
  this.tableName = null;
  this.alias = null;
  this.distinctFlag = false;
  this.fields = [];
  this.clauses = [];
  this.joins = [];
};

QueryBuilder.prototype.select = function() {
  this.operation = 'select';

  return this;
};

QueryBuilder.prototype.update = function() {
  this.operation = 'update';

  return this;
};

QueryBuilder.prototype.delete = function() {
  this.operation = 'delete';

  return this;
};

QueryBuilder.prototype.insert = function() {
  this.operation = 'insert';

  return this;
};

QueryBuilder.prototype.table = function(tableName) {
  if (this.operation !== 'update') {
    throw new Error('Invalid method');
  }

  this.tableName = tableName;

  return this;
};

QueryBuilder.prototype.into = function(tableName) {
  if (this.operation !== 'insert') {
    throw new Error('Invalid method');
  }

  this.tableName = tableName;

  return this;
};

QueryBuilder.prototype.from = function(tableName) {
  if (this.operation !== 'select') {
    throw new Error('Invalid method');
  }

  this.tableName = tableName;

  return this;
};

QueryBuilder.prototype.as = function(alias) {
  if (this.operation !== 'select') {
    throw new Error('Invalid method');
  }

  this.alias = alias;

  return this;
};


QueryBuilder.prototype.set = function(name, value, options) {
  this.fields.push({
    name: name,
    value: options && options.dontQuote ? knex.raw(value) : value
  });

  return this;
};

QueryBuilder.prototype.field = function(name) {
  this.fields.push({
    name: name
  });

  return this;
};

QueryBuilder.prototype.where = function() {
  this.clauses.push({
    criteria: arguments[0],
    values: Array.prototype.slice.call(arguments, 1)
  });

  return this;
};

QueryBuilder.prototype.distinct = function() {
  this.distinctFlag = true;

  return this;
};

QueryBuilder.prototype.join = function(tableName, alias, clause) {
  this.joins.push({
    tableName: tableName,
    alias: alias,
    clause: clause,
    type: 'inner'
  });

  return this;
};

QueryBuilder.prototype.left_join = function(tableName, alias, clause) {
  this.joins.push({
    tableName: tableName,
    alias: alias,
    clause: clause,
    type: 'left'
  });

  return this;
};

QueryBuilder.prototype.compile = function() {
  if (!this.operation) {
    throw new Error('No operation')
  }

  if (!this.tableName) {
    throw new Error('No table name');
  }

  if (this.operation === 'select') {
    var fieldNames = this.fields.map(function(field) {
      return field.name;
    });

    var query;

    if (this.distinctFlag) {
      query =  knex.select()
        .distinct(fieldNames)
        .from(this.tableName);
    } else {
      query = knex.select(fieldNames)
        .from(this.tableName);
    }

    if (this.alias) {
      query.as(this.alias);
    }

    this.joins.forEach(function(join) {
      if (join.type === 'left') {
        query.joinRaw(`left join ${join.tableName} ${join.alias ? `${join.alias}` : ''} on (${join.clause})`);
      } else {
        query.joinRaw(`inner join ${join.tableName} ${join.alias ? `${join.alias}` : ''} on (${join.clause})`);
      }
    });

    this.clauses.forEach(function(clause) {
      query.whereRaw('(' + clause.criteria + ')', clause.values);
    });

    return query;
  } else if (this.operation === 'insert') {
    var fieldValues = {};

    this.fields.forEach(function(field) {
      fieldValues[field.name] = field.value;
    });

    var query = knex.insert(fieldValues)
      .into(this.tableName);

    return query;
  } else if (this.operation === 'update') {
    var fieldValues = {};

    this.fields.forEach(function(field) {
      fieldValues[field.name] = field.value;
    });

    var query = knex.update(fieldValues)
      .table(this.tableName);

    this.clauses.forEach(function(clause) {
      query.whereRaw('(' + clause.criteria + ')', clause.values);
    });

    return query;
  } else if (this.operation === 'delete') {
      var query = knex.del(fieldValues)
      .from(this.tableName);

    this.clauses.forEach(function(clause) {
      query.whereRaw('(' + clause.criteria + ')', clause.values);
    });

    return query;
  }
};

QueryBuilder.prototype.toQuery = function() {
  return this.compile();
};

QueryBuilder.prototype.toRaw = function() {
  return knex.raw(this.compile());
};

QueryBuilder.prototype.toString = function() {
  return this.compile().toString();
};

module.exports = {
  select: function() {
    var queryBuilder = new QueryBuilder();

    return queryBuilder.select();
  },
  insert: function() {
    var queryBuilder = new QueryBuilder();

    return queryBuilder.insert();
  },
  update: function() {
    var queryBuilder = new QueryBuilder();

    return queryBuilder.update();
  },
  delete: function() {
    var queryBuilder = new QueryBuilder();

    return queryBuilder.delete();
  }
};
```

It is not complete by any means, but if anybody else wants to eject out of this package, feel free to make that a module, then replace every `squel.select()`, `squel.insert(), `squel.delete()`, and `squel.update()` call with its respective queryBuilder function. Drop in replacement. :)

Maintainer, I would highly suggest putting in big, large, red letters on the README for this package that this project offers 1/20th the performance of knex, for 1% syntax difference. I'd imagine that'd sway many people from using it, and prevent multiple people from building entire codebases, trusting their SQL query builder open source library package to be... performant.

**Comment by schmod** [comment]
Hey all.  

I just want to remind you all that Ramesh (who has a name that isn't "maintainer") has invested hundreds of hours of his time over the past 5 years _FOR FREE_ maintaining this library, with great documentation, an extensive feature-set, and rock-solid reliability.

As an open-source project, Squel belongs to the community, and if you have productive suggestions for improving the performance of this library, please feel free to submit them.  (I've personally provided several performance-related pull-requests to this project, all of which were promptly merged by @hiddentao)

As a query-generation library, a 20x performance gap, while possibly noticeable, likely isn't going to be the source of any bottlenecks in an application, given that network latency (and the database operation itself) is going to be several orders of magnitude slower than the process of generating the SQL itself.  

If knex fits your needs better than squel, go ahead and use it.  If you're using squel, but would like to see it improved, go ahead and submit a pull-request.  Don't dunk on other people's work, though –especially when you've benefitted from that work and have provided nothing but insults in return.

**Comment by brandonros** [comment]
I think you are missing the point. I am not undermining Ramesh's efforts. I
can not even imagine what it takes to put 5 years into a open source
project.

However, he is causing grief to others. I'm sorry to be the bearer of bad
and blunt news, and to come off as rude, but... yes... 20x performance gap
*did* slow down my app and *was* the main bottleneck. I felt very stupid
for trusting a open source library, only to profile why my CPU was pegged,
to find out that this library has glaring performance issues.

I bet Ramesh could fix the performance problems in a day. The fact that it
is ignored, and left for users of the library to find out after investing
hours, days, weeks into the project is not fair.

Unbiased, fair suggestion: disclose in the README that this package should
not be used if your app needs performance, and suggest knex instead.

**Comment by wirehead** [comment]
Unbiased, fair suggestion: Either submit code that you claim can be written in a day or stop abusing the maintainer, @brandonros.  There is nothing to be gained by being rude to people who owe you nothing.

**Comment by brandonros** [comment]
@wirehead No need. I wrote code to eject out of it and go to a library that doesn't put wild performance issues on the back burner for 2 years and not inform their users about them.

**Comment by hiddentao** [comment]
@wirehead @schmod Thanks for the kind words and support guys.

@brandonros I have worked on other things over the past couple of years, as priority over this. But honestly, I should get my ass into gear and do some work on this. I'm going to have a go at improving the performance. Knex will probably still remain more performant than Squel since it's much more actively developed. I might add something to the README to this effect as per your suggestion.

**Comment by brandonros** [comment]
Thanks Ramesh. Killer project. I've used it in so much stuff that when I
found out the performance wasn't the best... I was like... damn... HOW many
lines of code are affected? :P

Super easy to adapt and very clear/concise API. Every edge case I ran into,
you had me covered.

Thanks again!
